src/Compilers/Core/Portable/MetadataReader/PEModule.cs

```diff
@@ -3613,23 +3622,20 @@ private ConstantValue GetConstantValueOrThrow(ConstantHandle handle)
 
             if (ignoreCase)
             {
-                // This linear search is not the optimal way to use a hashmap, but we should only use
-                // this functionality when computing diagnostics.  Note
-                // that we can't store the map case-insensitively, since real metadata name
-                // lookup has to remain case sensitive.
-                foreach (var pair in _lazyForwardedTypesToAssemblyIndexMap)
+                // We should only use this functionality when computing diagnostics, so we lazily construct
+                // a case-insensitive map when necessary. Note that we can't store the original map
+                // case-insensitively, since real metadata name lookup has to remain case sensitive.
+                ensureCaseInsensitiveDictionary();
+
+                if (_lazyCaseInsensitiveForwardedTypesToAssemblyIndexMap.TryGetValue(fullName, out var value))
                 {
-                    if (string.Equals(pair.Key, fullName, StringComparison.OrdinalIgnoreCase))
-                    {
-                        matchedName = pair.Key;
-                        return pair.Value;
-                    }
+                    matchedName = fullName;
```

Comment:
>matchedName = fullName [](http://example.com/codeflow?start=20&length=22)

Since ```matchedName``` parameter is meaningless now, let's remove it and follow up at the call sites accordingly.

Reply:
> Since `matchedName` parameter is meaningless now, let's remove it and follow up at the call sites accordingly.

Seems reasonable, @333fred?

Reply:
@ToddGrun Aleksey and I were working offline on determining that we were wrong about the observability of this change 🙂.